### PR TITLE
transact: support insert with uuid

### PR DESCRIPTION
**Pull request by hunchback**

### CODE REVIEW

File: pkg/ovsdb/transact.go [hunk 1]
```diff
@@ -374,6 +375,20 @@ func NewCondition(tableSchema *libovsdb.TableSchema, condition []interface{}) (*
 	}
 
 	value := condition[2]
+
+	/* replace named-uuid with uuid */
+	switch namedUUID := value.(type) {
```
Comment (roytman) [inline pkg/ovsdb/transact.go:380]: namedUUID cannot be part of a condition

File: pkg/ovsdb/transact.go [hunk 2]
```diff
@@ -648,14 +663,14 @@ func doesWhereContainCondTypeUUID(tableSchema *libovsdb.TableSchema, where []int
...
-func isRowSelectedByWhere(tableSchema *libovsdb.TableSchema, row *map[string]interface{}, where []interface{}) (bool, error) {
+func isRowSelectedByWhere(tableSchema *libovsdb.TableSchema, mapUUID MapUUID, row *map[string]interface{}, where []interface{}) (bool, error) {
 	for _, c := range where {
 		cond, ok := c.([]interface{})
 		if !ok {
 			klog.Errorf("Failed to convert condition value: %v", c)
 			return false, errors.New(E_INTERNAL_ERROR)
 		}
-		ok, err := isRowSelectedByCond(tableSchema, row, cond)
+		ok, err := isRowSelectedByCond(tableSchema, mapUUID, row, cond)
```
Comment (roytman) [inline pkg/ovsdb/transact.go:673]: you cannot select Row by namedUUID

File: pkg/ovsdb/transact.go [hunk 3]
```diff
@@ -1049,10 +1069,14 @@ func doInsert(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.O
...
-	if ovsOp.UUID != nil {
-		key.UUID = ovsOp.UUID.GoUUID
+	var uuid string
+	if ovsOp.UUID == nil {
+		uuid = common.GenerateUUID()
+	} else {
+		uuid = ovsOp.UUID.GoUUID
 	}
+
+	key := common.GenerateDataKey(txn.request.DBName, ovsOp.Table, uuid)
```
Comment (roytman) [inline pkg/ovsdb/transact.go:1079]: it should be NewDataKey

**Review by roytman — CHANGES_REQUESTED** [review]
UUIDNamed used in at least 2 operations in the same transaction:

- The Insert, were you create a new row with the a generate UUID which is  linked to the UUIDnamed.
- Another Update operation, where you can update columns of the existing row with the references to the UUID of the created ROW

File: pkg/ovsdb/transact.go [hunk 4]
```diff
@@ -1032,11 +1047,16 @@ func preInsert(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.
...
-		if ovsOp.UUIDName != "" && (*row)[COL_UUIDNAME] == ovsOp.UUIDName {
+	if ovsOp.UUIDName != "" {
+		if _, ok := txn.mapUUID[ovsOp.UUIDName]; ok {
 			klog.Errorf("Duplicate uuid-name: %s", ovsOp.UUIDName)
 			return errors.New(E_DUP_UUIDNAME)
 		}
+		uuid := common.GenerateUUID()
+		txn.mapUUID[ovsOp.UUIDName] = uuid
+	}
+
```
Comment (roytman) [inline pkg/ovsdb/transact.go:1058]: Where do you handling the option when a user specifies a required UUID for the new row

File: pkg/ovsdb/transact.go (hunk 3 above)
Comment (hunchback) [reply pkg/ovsdb/transact.go]: using NewDataKey now

File: pkg/ovsdb/transact.go (hunk 4 above)
Comment (hunchback) [reply pkg/ovsdb/transact.go:1106]: fixed, now handling above

File: pkg/common/key.go [hunk 5]
```diff
@@ -82,9 +82,14 @@ func (k *Key) ToTableKey() Key {
...
 
+/* GenerateUUID generate RFC4122 UUID */
+func GenerateUUID() string {
+	return guuid.NewString()
+}
+
 // Creates a new data key with a generated uuid
-func GenerateDataKey(dbName, tableName string) Key {
-	return NewDataKey(dbName, tableName, guuid.NewString()) /* generate RFC4122 UUID */
+func GenerateDataKey(dbName, tableName, uuid string) Key {
```
[the remark this answers is not in the document]
Comment (hunchback) [reply pkg/common/key.go]: ok using NewDataKey

File: pkg/ovsdb/transact.go (hunk 1 above)
Comment (hunchback) [reply pkg/ovsdb/transact.go]: removed

File: pkg/ovsdb/transact.go (hunk 2 above)
Comment (hunchback) [reply pkg/ovsdb/transact.go]: ok

**Pull request merged**